go/vt/vtgate/planbuilder/builder.go

```diff
@@ -216,3 +219,111 @@ func buildVSchemaDDLPlan(stmt *sqlparser.AlterVschema, vschema ContextVSchema) (
 		AlterVschemaDDL: stmt,
 	}, nil
 }
+
+func buildFlushPlan(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	if len(stmt.TableNames) == 0 {
+		return buildFlushOptions(stmt, vschema)
+	}
+	return buildFlushTables(stmt, vschema)
+}
+
+func buildFlushOptions(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	destination, keyspace, _, err := vschema.TargetDestination("")
+	if err != nil {
+		return nil, err
+	}
+	return &engine.Send{
+		Keyspace:          keyspace,
+		TargetDestination: destination,
+		Query:             sqlparser.String(stmt),
+		IsDML:             false,
+		SingleShardOnly:   false,
+	}, nil
+}
+
+func buildFlushTables(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	type sendDest struct {
+		ks   *vindexes.Keyspace
+		dest key.Destination
+	}
+
+	flushStatements := map[sendDest]*sqlparser.Flush{}
+	for i, tab := range stmt.TableNames {
+		var destinationTab key.Destination
+		var keyspaceTab *vindexes.Keyspace
+		var table *vindexes.Table
+		var err error
+		table, _, _, _, destinationTab, err = vschema.FindTableOrVindex(tab)
+
+		if err != nil {
+			_, isNotFound := err.(vindexes.NotFoundError)
+			if !isNotFound {
+				return nil, err
+			}
+		}
+		if table == nil {
+			destinationTab, keyspaceTab, _, err = vschema.TargetDestination(tab.Qualifier.String())
+			if err != nil {
+				return nil, err
+			}
+			stmt.TableNames[i] = sqlparser.TableName{
+				Name: tab.Name,
+			}
+		} else {
+			keyspaceTab = table.Keyspace
+			stmt.TableNames[i] = sqlparser.TableName{
```

Comment:
we do not have to do this for flush statements. This logic make sense for ddl statements but not needed for flush statement.
checking in vschema should be good enough

Reply:
why is that?